zebra-network/src/peer_set/initialize.rs

```diff
@@ -257,14 +257,34 @@ where
             }
             Err((addr, ref e)) => {
                 handshake_error_total += 1;
+
                 // this is verbose, but it's better than just hanging with no output when there are errors
-                info!(
-                    ?handshake_success_total,
-                    ?handshake_error_total,
-                    ?addr,
-                    ?e,
-                    "an initial peer connection failed"
-                );
+                let mut expected_error = false;
+                if let Some(io_error) = e.downcast_ref::<tokio::io::Error>() {
+                    // Some systems only have IPv4, or only have IPv6,
+                    // so these errors are not particularly interesting.
+                    if io_error.kind() == tokio::io::ErrorKind::AddrNotAvailable {
+                        expected_error = true;
+                    }
+                }
+
+                if expected_error {
+                    debug!(
+                        successes = ?handshake_success_total,
+                        errors = ?handshake_error_total,
+                        ?addr,
+                        ?e,
+                        "an initial peer connection failed"
+                    );
+                } else {
+                    info!(
+                        successes = ?handshake_success_total,
+                        errors = ?handshake_error_total,
+                        ?addr,
+                        %e,
+                        "an initial peer connection failed"
+                    );
+                }
```

Comment:
Rant: unfortunately `tracing` doesn't let us do something like:

```suggestion
                let mut log_level = Level::INFO;
                if let Some(io_error) = e.downcast_ref::<tokio::io::Error>() {
                    // Some systems only have IPv4, or only have IPv6,
                    // so these errors are not particularly interesting.
                    if io_error.kind() == tokio::io::ErrorKind::AddrNotAvailable {
                        log_level = Level::DEBUG;
                    }
                }

                event!(
                    log_level,
                    successes = ?handshake_success_total,
                    errors = ?handshake_error_total,
                    ?addr,
                    ?e,
                    "an initial peer connection failed"
                );
```

because the `event!` macro seems to create a `static` variable that uses `log_level` :(

Reply:
Yep, this is a deliberate optimisation choice.
(It improves branch prediction, and sometimes lets Rust optimise out the log entirely.)